### ISSUE TITLE
Supertype of all arrays

### ISSUE THREAD

**Issue by askeksa-google** [WebAssembly/gc#259]
In `dart2wasm`, there are two different representations for strings - `OneBytetring` (containing Latin-1) and `TwoByteString` (containing WTF-16). Each is a wrapper struct containing some fields present in all Dart objects (class id, identity hash) plus a Wasm array. The element type of the array is `i8` and `i16`, respectively.

The `.length` getter on a string is a virtual call, since it will need to dispatch to specific code for reading the length of the array of `i8` or `i16`. This seems a bit silly, since the underlying engine likely has a partly uniform representation of arrays where the length could be accessed in the same place regardless of the element type.

This gave me an idea: we could have an `arrayref` type which is a subtype of `dataref` and a supertype of all arrays. The only operation permitted on an `arrayref` (on top of the ones for `dataref`) is `array.len`.

With this, the common string supertype of the two string types could declare an immutable field of type `arrayref` for the contents array, which would then be specialized in each subclass. The `.length` getter could simply execute an `array.len` on that field.

WDYT?

**Comment by titzer** [comment]
This sounds like a good idea to me. In Java class files, the `arraylength` bytecode will work on any type of array.

**Comment by rossberg** [comment]
This makes sense to me. But I assume we'd also need accompanying instructions like `ref.is_array` and `ref.as_array` to actually get values of this type.

**Comment by askeksa-google** [comment]
The `arrayref` type is useful even without these instructions (via implicit upcasts and field covariance), but it makes sense to add them for consistency.

**Comment by manoskouk** [comment]
To make this work we will have to remove the immediate array index from `array.len`.

**Comment by manoskouk** [comment]
Update: We just implemented `arrayref` in V8 including `ref.is_array`, `ref.as_array`, `br_on_array` and `br_on_non_array`. The opcodes and details can be found [here](https://docs.google.com/document/d/1DklC3qVuOdLHSXB5UXghM_syCh-4cMinQ50ICiXnK3Q/). Notably, we now ignore the type index immediate for `array.len` so that it can work on values of type `anyref`. For backwards compatibility, we did not drop the immediate yet.
